src/fiber/fiber.ml

```diff
@@ -546,17 +546,18 @@ module Mvar = struct
         EC.safe_run_k k v;
         K.run w ())
 
-  let write t x k =
+  let write t x =
```

Comment:
A similar change should perhaps be done to `read`.

Here's a piece of trivia: who will get "first", and who will get "second"?

```ocaml
let mvar = Mvar.create_full "first" in
let w () = Mvar.write mvar "second" in
let f1 = Mvar.read mvar >>| print_endline in
let f2 = Mvar.read mvar >>| print_endline in
Fiber.fork_and_join_unit w (fun () -> f2 >>> f1)
```

I think the result is counter intuitive.

Reply:
I expect:

- `f2` gets `first`
- `f1` gets `second`

Because writing `let x = <some fiber> in` has no effect in the fiber world. Fibers are not like Lwt or Async; a fiber is not started until it is attached to the toplevel `run` via a series of bind.

In particular, this diff looks dodgy to me. The previous code was doing the side effects inside the fiber, which is the right thing to do.

Reply:
I see. I always interpreted these actions as enqueue my read and then run my pure computation when it is ready. I'm not sure how to make `ignore (Fiber.Pool.task pool ~f)` work then. We need to be able to enqueue ourselves without being stuck in a fiber for RPC.

Reply:
`ignore (Fiber.Pool.task pool ~f)` definitely shouldn't work. For the RPC, as long as submitting the task doesn't wait for it to be executed, then you should be good. You just need to propagate the monad enough.

Reply:
The way to think about of `Pool.task pool ~f` is that it represents the action of submitting `f` to the pool. But until the fiber is effectively executed, it has no effect at all. Which is generally the right way to think about monads.

Reply:
> We need to be able to enqueue ourselves without being stuck in a fiber for RPC.

This is what I wanted to avoid. It's probably not going to be so bad to propagate the monad in this case. Build system already uses it liberally.